src/components/themed/ExchangedFlipInput2.tsx

```diff
@@ -0,0 +1,208 @@
+import { div, log10, mul, toFixed } from 'biggystring'
+import { EdgeCurrencyWallet } from 'edge-core-js'
+import React, { useState } from 'react'
+import { ActivityIndicator, Platform, TouchableOpacity } from 'react-native'
+
+import { useDisplayDenom } from '../../hooks/useDisplayDenom'
+import { useExchangeDenom } from '../../hooks/useExchangeDenom'
+import { useHandler } from '../../hooks/useHandler'
+import { useWatch } from '../../hooks/useWatch'
+import { useSelector } from '../../types/reactRedux'
+import { getCurrencyCode } from '../../util/CurrencyInfoHelpers'
+import { DECIMAL_PRECISION, getDenomFromIsoCode, maxPrimaryCurrencyConversionDecimals, precisionAdjust } from '../../util/utils'
+import { CryptoIcon } from '../icons/CryptoIcon'
+import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
+import { EdgeText } from './EdgeText'
+import { FlipInput2, FlipInputFieldInfo, FlipInputGetMethodsResponse } from './FlipInput2'
+import { RightChevronButton } from './ThemedButtons'
+
+export type ExchangeFlipInputFields = 'fiat' | 'crypto'
+
+export interface ExchangedFlipInputGetMethodsResponse {
+  setAmount: (field: ExchangeFlipInputFields, value: string) => void
+}
+
+export type ExchangedFlipInputAmounts = {
+  exchangeAmount: string
+  nativeAmount: string
+  fiatAmount: string
+  fieldChanged: 'fiat' | 'crypto'
+}
+
+export type ExchangedFlipInputProps = {
+  walletId: string
+  tokenId?: string
+  startNativeAmount?: string
+  headerText: string
+  headerCallback?: () => void
+  onAmountChanged: (amounts: ExchangedFlipInputAmounts) => unknown
+  getMethods?: (methods: ExchangedFlipInputGetMethodsResponse) => void
```

Comment:
I figured this would be the biggest area of debate. While refs are native to React, their syntax is awkward and a regular developer wouldn't understand them without further doc research (I screwed them up myself). The getMethods callback is exactly the API used in the FiatPlugins that you reviewed and its the API we all liked for the SafeSlider. It accomplishes exactly the same thing, provides better security (can't call methods that the component doesn't give access to), and needs no additional understanding of React syntax, and is easier to chain deeper into nested child components. With this, haven't found an advantage to refs.

Reply:
I 100% disagree. Instead of using the documented thing that the library provides, we are going off and inventing our own thing. This is way worse for productivity, since now you need to learn our undocumented thing, plus the well-documented built-in thing which exists in the app anyhow.

The refs syntax *was* awkward in the class component days, but now it is awesome. Just do `const textInput = useRef<TextInput>(null)`, and attach it with `<TextInput ref={textInput} />`. Now you just call `textInput.current?.whateverMethod()` as needed. You don't need to write any callbacks or manually manage state as you do in the `getMethods` approach - React automatically updates the ref.

On the implementation side, you have precise control over what methods you expose. You pass this interface definition to `forwardRef<MyMethods, Props>`, and that's all you get. No leaks. Then, instead of using a `useEffect` to kick off the `getMethods` callback, you call `useImperativeHandle(ref, () => ({...methods}))`, which does basically the same thing.

The only real advantage you have mentioned is that `getMethods` is easier to thread up the stack, which is true. It doesn't apply in this case, though - the exchanged flip input exposes different methods than the inner flip input does, since it provides a higher-level API.

Generally speaking, directly exposing the methods from a nested child suggests a bad design, since it's leaking internal implementation to the outside. So, the fact that refs require a tiny bit of effort to forward may actually be an advantage, since you aren't going to leak internals without thinking first.

I think the real solution here is to educate our developers - if people don't know how to use the framework our app is built on, we should help them get up to speed. Avoiding the issue isn't a professional approach.

Reply:
Per sidebar convo. Commit https://github.com/EdgeApp/edge-react-gui/pull/3712/commits/fbca49c746581c207e6087f3c8f144c277ce31c3 converted to using refs